src/commands/test/runner.py

```diff
@@ -196,29 +184,66 @@ async def invoke_test_function(self, function_name: str):
 
         func = getattr(self.test_contract, function_name)
 
-        # TODO: Improve stacktrace
+        # pylint: disable=too-many-nested-blocks
         try:
-            call_result = await func().invoke()
-            if self._expected_error is not None:
-                raise MissingExceptException(
-                    f"Expected an exception matching the following error: {self._expected_error}"
-                )
-            return call_result
-
-        except StarkException as ex:
-            if self._expected_error:
-                if not self._expected_error.match(ex):
-                    raise ExceptMismatchException(
-                        expected_name=self._expected_error.name,
-                        expected_message=self._expected_error.message,
-                        received=ex,
+            try:
+                try:
+                    call_result = await func().invoke()
+                    for listener in self._test_finished_listener_map.values():
+                        if listener:
+                            listener()
+
+                    if self._expected_error is not None:
+                        raise StandardReportedException(
+                            f"Expected an exception matching the following error:\n{self._expected_error}"
+                        )
+                    return call_result
+                # catch StarkNet exceptions and wrap them with RevertableException
+                except StarkException as ex:
+                    raise RevertableException(
+                        error_type=ex.code.name,
+                        error_message=extract_core_info_from_stark_ex_message(
+                            ex.message
+                        ),
+                        exception=ex,
                     ) from ex
-            else:
-                raise StarkReportedException(ex) from ex
+            # catch a RevertableException and rethrow the original_exception if the exception is not expected
+            except RevertableException as ex:
+                if self._expected_error:
+                    if not self._expected_error.match(ex):
+                        raise ExpectedRevertMismatchException(
+                            expected=self._expected_error,
+                            received=ex,
+                        ) from ex
+                else:
+                    if ex.original_exception:
+                        raise ex.original_exception
```

Comment:
Do we need this extra `try-except` layer? Can't we just raise `StandardReportedException` here? Would be much much easier to comprehend.